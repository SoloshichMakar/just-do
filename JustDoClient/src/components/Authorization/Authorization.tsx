import React from "react";
import { Snackbar } from "@material-ui/core";
import JustDoLogo from "../../picture/JustDoLogo.png";
import JustDoComponent from "../../picture/JustDoComponent.png";
import FormControl from "@material-ui/core/FormControl";
import Input from "@material-ui/core/Input";
import {
  emailHelperMessage,
  passwordHelperMessage,
  validateEmail,
  validatePassword,
} from "../../utils/validateUserDataFunction";
import { NavLink, useHistory } from "react-router-dom";
import { gql } from "apollo-boost";
import { useMutation } from "@apollo/react-hooks";
import { Alert } from "@material-ui/lab";
import { ColorButton, CustomLink } from "../style/CustomElements";

interface IAuthorization {
  email: string;
  password: string;
  message: string;
  isCreated: boolean;
  success: false;
  formClear: Function;
  authenticated: Function;
  errorMessage: Function;
  clearMessage: Function;
  textEmailChange: Function;
  textPasswordChange: Function;
}

let emailIsEdit = false;
let passwordIsEdit = false;

const USER_LOGIN = gql`
  mutation Login($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      user {
        id
        email
      }
      token
    }
  }
`;

const Authorization: React.FC<IAuthorization> = ({
  email,
  password,
  message,
  success,
  isCreated,
  errorMessage,
  clearMessage,
  formClear,
  authenticated,
  textEmailChange,
  textPasswordChange,
}) => {
  const [login] = useMutation(USER_LOGIN);
  const history = useHistory();

  function currentTextEmailChange(e: React.ChangeEvent<HTMLInputElement>) {
    emailIsEdit = true;
    const getTextAreaValue = e.target.value;
    textEmailChange(getTextAreaValue);
  }

  function currentTextPasswordChange(e: React.ChangeEvent<HTMLInputElement>) {
    passwordIsEdit = true;
    const getTextAreaValue = e.target.value;
    textPasswordChange(getTextAreaValue);
  }

  function redirectToSighUp() {
    formClear();
    emailIsEdit = false;
    passwordIsEdit = false;
  }

  async function userAuthorization() {
    try {
      if (
        !validateEmail(emailIsEdit, email) &&
        emailIsEdit &&
        !validatePassword(passwordIsEdit, password) &&
        passwordIsEdit
      ) {
        const response = await login({
          variables: { email: email, password: password },
        });
        localStorage.setItem("token", response.data.login.token);
        authenticated();
        emailIsEdit = false;
        passwordIsEdit = false;
        history.push("/");
      }
    } catch (error) {
      errorMessage();
    }
  }

  const handleErrorClose = (event?: React.SyntheticEvent, reason?: string) => {
    if (reason === "clickaway") {
      return;
    }
    clearMessage();
  };

  const handleClose = (event?: React.SyntheticEvent, reason?: string) => {
    if (reason === "clickaway") {
      return;
    }
    clearMessage();
  };

  console.log(success);
  return (
    <div id="registration_main">
      <Snackbar
        open={message.length !== 0 && !success}
        autoHideDuration={10000}
        onClose={handleErrorClose}
      >
        <Alert onClose={handleErrorClose} severity="error">
          {message}
        </Alert>
      </Snackbar>
      <Snackbar
        open={isCreated || success}
        autoHideDuration={10000}
        onClose={handleClose}
      >
        <Alert onClose={handleClose} severity="success">
          {message}
        </Alert>
      </Snackbar>
      <div id="registration_main__logo_images">
        <img id="logo" src={JustDoLogo} />
        <img src={JustDoComponent} id="background_image" />
      </div>
      <div id="registration_main__components">
        <h1>Sign in</h1>
        <div id="components__input_components">
          <div className="input_components__text_input">
            <FormControl id="textInput__email">
              <Input
                id="email_input"
                error={validateEmail(emailIsEdit, email)}
                type="text"
                value={email}
                onChange={currentTextEmailChange}
                placeholder="E-mail"
              />
              <div id="errorMessage">
                {emailHelperMessage(emailIsEdit, email)}
              </div>
            </FormControl>
          </div>

          <div className="input_components__text_input">
            <FormControl id="textInput__passwordForm">
              <Input
                id="password_input"
                error={validatePassword(passwordIsEdit, password)}
                type={"password"}
                value={password}
                placeholder="Password"
                onChange={currentTextPasswordChange}
              />
              <div id="errorMessage">
                {passwordHelperMessage(passwordIsEdit, password)}
              </div>
            </FormControl>
          </div>
        </div>

        <div id="input_components__color_button">
          <div id="input_components__password_restore">
            <NavLink to="/password_restore">
              <CustomLink href="#" id="forgot_password">
                Forgot password?
              </CustomLink>
            </NavLink>
          </div>

          <ColorButton
            id="color_button"
            variant="contained"
            color="primary"
            onClick={() => userAuthorization()}
          >
            Sign in
          </ColorButton>
        </div>
        <div id="input_components__footer__sign_in">
          <NavLink to="/registration">
            <CustomLink
              href="/registration"
              id="sign_in__link"
              onClick={() => redirectToSighUp()}
            >
              Sign Up
            </CustomLink>
          </NavLink>
        </div>
        <div id="input_components__footer__agreements">
          <div>By accessing your account, you agree to our</div>
          <div id="agreements__links">
            <NavLink to="/#">
              <CustomLink
                id="item__terms_conditions"
                className="agreements__links__item"
                href="#"
              >
                Terms conditions
              </CustomLink>
            </NavLink>
            <div className="agreements__links__item"> and </div>
            <NavLink to="/#">
              <CustomLink
                id="item__privacy_policy"
                className="agreements__links__item"
              >
                Privacy Policy
              </CustomLink>
            </NavLink>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Authorization;
