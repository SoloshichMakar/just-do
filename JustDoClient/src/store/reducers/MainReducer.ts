import RegistrationReducer from "./RegistrationReducer";
import AuthorizationReducer from "./AuthorizationReducer";
import PasswordRestoreReducer from "./PasswordRestoreReducer";
import ForgotPasswordReducer from "./ForgotPasswordReducer";

import { combineReducers } from "redux";

const MainReducer = combineReducers({
  RegistrationReducer,
  AuthorizationReducer,
  PasswordRestoreReducer,
  ForgotPasswordReducer,
});

export default MainReducer;
