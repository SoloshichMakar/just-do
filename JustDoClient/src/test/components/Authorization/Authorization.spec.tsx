import Authorization from "../../../components/Authorization/Authorization";
import React from "react";
import renderer from "react-test-renderer";
import { createStore } from "redux";
import MainReducer from "../../../store/reducers/MainReducer";
import { Provider } from "react-redux";
import ApolloClient from "apollo-boost";
import { ApolloProvider } from "react-apollo";
import {BrowserRouter} from "react-router-dom";


const props = {
  email: "",
  password: "",
  message: "",
  isCreated: false,
  success: false,
  formClear: Function,
  authenticated: Function,
  errorMessage: Function,
  clearMessage: Function,
  textEmailChange: Function,
  textPasswordChange: Function,
};

const client = new ApolloClient({
  uri: ``,
});
const store = createStore(MainReducer);

describe("Authorization snapshot", () => {
  it("Authorization render correctly", () => {

    const tree = renderer
      .create(
        <ApolloProvider client={client}>
          <Provider store={store}>
            <BrowserRouter>
            <Authorization
              email={props.email}
              password={props.password}
              authenticated={props.authenticated}
              clearMessage={props.clearMessage}
              errorMessage={props.errorMessage}
              formClear={props.formClear}
              isCreated={props.isCreated}
              message={props.message}
              success={props.success}
              textEmailChange={props.textEmailChange}
              textPasswordChange={props.textPasswordChange}
            />
            </BrowserRouter>
          </Provider>
        </ApolloProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
