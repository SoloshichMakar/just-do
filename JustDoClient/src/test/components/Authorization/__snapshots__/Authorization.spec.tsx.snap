// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization snapshot Authorization render correctly 1`] = `
<div
  id="registration_main"
>
  <div
    id="registration_main__logo_images"
  >
    <img
      id="logo"
      src="JustDoLogo.png"
    />
    <img
      id="background_image"
      src="JustDoComponent.png"
    />
  </div>
  <div
    id="registration_main__components"
  >
    <h1>
      Sign in
    </h1>
    <div
      id="components__input_components"
    >
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__email"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="email_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              required={false}
              type="text"
              value=""
            />
          </div>
          <div
            id="errorMessage"
          />
        </div>
      </div>
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__passwordForm"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="password_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              required={false}
              type="password"
              value=""
            />
          </div>
          <div
            id="errorMessage"
          />
        </div>
      </div>
    </div>
    <div
      id="input_components__color_button"
    >
      <div
        id="input_components__password_restore"
      >
        <a
          className="MuiTypography-root MuiLink-root WithStyles(ForwardRef(Link))-root-44 MuiLink-underlineHover MuiTypography-colorPrimary"
          href="/password_restore"
          id="forgot_password"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Forgot password?
        </a>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-81 MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        id="color_button"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Sign in
        </span>
      </button>
    </div>
    <div
      id="input_components__footer__sign_in"
    >
      <a
        className="MuiTypography-root MuiLink-root WithStyles(ForwardRef(Link))-root-44 MuiLink-underlineHover MuiTypography-colorPrimary"
        href="/registration"
        id="sign_in__link"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        Sign Up
      </a>
    </div>
    <div
      id="input_components__footer__agreements"
    >
      <div>
        By accessing your account, you agree to our
      </div>
      <div
        id="agreements__links"
      >
        <a
          className="MuiTypography-root MuiLink-root WithStyles(ForwardRef(Link))-root-44 MuiLink-underlineHover agreements__links__item MuiTypography-colorPrimary"
          href="/#"
          id="item__terms_conditions"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Terms conditions
        </a>
        <div
          className="agreements__links__item"
        >
           and 
        </div>
        <a
          className="MuiTypography-root MuiLink-root WithStyles(ForwardRef(Link))-root-44 MuiLink-underlineHover agreements__links__item MuiTypography-colorPrimary"
          href="/#"
          id="item__privacy_policy"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Privacy Policy
        </a>
      </div>
    </div>
  </div>
</div>
`;
